### PR TITLE
Russian translation added.

### PR DESCRIPTION
**Pull request by DJm00n**
I think there is some work needed in Solar for better localization. You need to use [gettext.ngettext](http://docs.python.org/3/library/gettext.html#gettext.ngettext) function for plural forms and do _NOT_ concatenate strings like this:

<pre>
_UNIFYING_RECEIVER_TEXT = (
                gettext.ngettext("No device paired") + '.\n\n<small>' + _("Up to %d devices can be paired to this receiver") + '.</small>',
                '%d ' + _("paired devices") + '\n\n<small>' + _("Up to %d devices can be paired to this receiver") + '.</small>',
</pre>

Instead write something like this:

<pre>
_UNIFYING_RECEIVER_TEXT = (
                _("No device paired.\n\n<small>"Up to %d devices can be paired to this receiver.</small>",
                _("%d paired devices.\n\n<small>Up to %d devices can be paired to this receiver.</small>",
</pre>

See http://techbase.kde.org/Development/Tutorials/Localization/i18n_Mistakes for details.

---

I'll try to do some work on this if I will have free time.

---

Update: Added some example localization fixes. Trying not to break anything. :)

### CODE REVIEW

**Comment by Lekensteyn** [comment]
Hey Dimitriy, there are some conflicts. Can you rebase it on top of the latest master?

You are absolutely right that translatable texts should not be split up. Would you also be OK with syncing the .pot file with the current sources?

**Comment by DJm00n** [comment]
@Lekensteyn I rebased this PR on current master. Try to merge now.

File: lib/logitech_receiver/status.py [hunk 1]
```diff
@@ -95,9 +96,8 @@ def __init__(self, receiver, changed_callback):
 
 	def __str__(self):
 		count = len(self._receiver)
-		return (_("No paired devices.") if count == 0 else
-				_("1 paired device.") if count == 1 else
-				(str(count) + _(" paired devices.")))
+		return (gettext.gettext("No paired devices.") if count == 0 else
```
Comment (Lekensteyn) [inline lib/logitech_receiver/status.py]: `_` is an alias for `gettext.gettext` (see `lib/solaar/i18n.py`), can you use that?

maybe it is an idea to export `ngettext` in i18n.py so you do not have to import gettext here.

File: lib/solaar/ui/about.py [hunk 2]
```diff
@@ -70,6 +70,7 @@ def _create():
 					'Adrian Piotrowicz (polski)',
 					'Daniel Pavel (română)',
 					'Daniel Zippert, Emelie Snecker (svensk)',
+					'Dimitriy Ryazantcev (Russian)'
```
Comment (Lekensteyn) [inline lib/solaar/ui/about.py]: Consider adding a trailing comma here

File: lib/solaar/ui/pair_window.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
 
 
 from solaar.i18n import _
+import gettext
```
Comment (Lekensteyn) [inline lib/solaar/ui/pair_window.py]: no need to import this

File: lib/solaar/ui/window.py [hunk 4]
```diff
@@ -538,11 +522,11 @@ def _details_items(device, read_all=False):
 				for fw in list(device.firmware):
 					yield ('  ' + _(str(fw.kind)), (fw.name + ' ' + fw.version).strip())
 			elif device.kind is None or device.online:
-				yield ('  %s' % _("Firmware"), '...')
+				yield (_("  Firmware"), '...')
```
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: Can you undo this change? The space is for formatting purposes, the original translation should be OK.

File: lib/solaar/ui/window.py [hunk 5]
```diff
@@ -538,11 +522,11 @@ def _details_items(device, read_all=False):
...
 
 			flag_bits = device.status.get(_K.NOTIFICATION_FLAGS)
 			if flag_bits is not None:
-				flag_names = ('(%s)' % _("none"),) if flag_bits == 0 else _hidpp10.NOTIFICATION_FLAG.flag_names(flag_bits)
+				flag_names = _("(none)") if flag_bits == 0 else _hidpp10.NOTIFICATION_FLAG.flag_names(flag_bits)
```
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: Same here, the format is normally always `(X)`, so I guess it makes more sense to make `X` a translatable string

File: lib/solaar/ui/window.py [hunk 6]
```diff
@@ -578,16 +562,15 @@ def _update_receiver_panel(receiver, panel, buttons, full=False):
...
-			panel._count.set_markup(_UNIFYING_RECEIVER_TEXT[0] % receiver.max_devices)
-		else:
-			panel._count.set_markup(_UNIFYING_RECEIVER_TEXT[1] % (devices_count, receiver.max_devices))
-	else:
-		if devices_count == 0:
-			panel._count.set_markup(_NANO_RECEIVER_TEXT[0])
-		else:
-			panel._count.set_markup(_NANO_RECEIVER_TEXT[1])
+	
+	paired_text = gettext.gettext('No device paired.') if devices_count == 0 else gettext.ngettext('One paired device.', '%(devices_count)s paired devices.', devices_count) % { 'devices_count': devices_count }
```
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: Use `_` instead of `gettext.gettext`?

File: lib/solaar/ui/window.py [hunk 7]
```diff
@@ -578,16 +562,15 @@ def _update_receiver_panel(receiver, panel, buttons, full=False):
...
+		
+	if(receiver.max_devices > 0):
+		paired_text += gettext.ngettext('\n\n<small>Up to %(max_devices_count)s device can be paired to this receiver.</small>', '\n\n<small>Up to %(max_devices_count)s devices can be paired to this receiver.</small>', receiver.max_devices) % { 'max_devices_count': receiver.max_devices }
```
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: Please separate formatting from translations, translators should not have to see HTML markup I think

File: lib/solaar/ui/window.py [hunk 8]
```diff
@@ -625,22 +608,22 @@ def _update_device_panel(device, panel, buttons, full=False):
 		panel._battery._icon.set_sensitive(False)
 		panel._battery._icon.set_from_icon_name(icon_name, _INFO_ICON_SIZE)
 		panel._battery._text.set_sensitive(True)
-		panel._battery._text.set_markup('<small>%s</small>' % _("unknown"))
+		panel._battery._text.set_markup(_("<small>unknown</small>"))
```
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: same here, separate formatting from translations

File: lib/solaar/ui/window.py [hunk 9]
```diff
@@ -625,22 +608,22 @@ def _update_device_panel(device, panel, buttons, full=False):
...
 	else:
 		charging = device.status.get(_K.BATTERY_CHARGING)
 		icon_name = _icons.battery(battery_level, charging)
 		panel._battery._icon.set_from_icon_name(icon_name, _INFO_ICON_SIZE)
 		panel._battery._icon.set_sensitive(True)
 
 		if isinstance(battery_level, _NamedInt):
-			text = _(str(battery_level))
+			text = _("%(battery_level)s") % { 'battery_level': _(str(battery_level)) }
```
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: Any reason why this is translated? It is just a number. Same question below.

File: lib/solaar/ui/window.py (hunk 4 above)
Comment (DJm00n) [inline lib/solaar/ui/window.py]: ok

File: lib/solaar/ui/window.py (hunk 5 above)
Comment (DJm00n) [inline lib/solaar/ui/window.py]: ok

File: lib/solaar/ui/window.py (hunk 7 above)
Comment (DJm00n) [inline lib/solaar/ui/window.py]: html is ok in translations:
HTML markup, however, is common enough that it’s probably ok to use in translatable strings. But please bear in mind that the GNU gettext tools don’t verify that the translations are well-formed HTML.
https://www.gnu.org/software/gettext/manual/gettext.html

File: lib/solaar/ui/about.py (hunk 2 above)
Comment (DJm00n) [inline lib/solaar/ui/about.py]: ok

File: lib/solaar/ui/pair_window.py (hunk 3 above)
Comment (DJm00n) [inline lib/solaar/ui/pair_window.py]: ok

File: lib/solaar/ui/window.py (hunk 9 above)
Comment (DJm00n) [inline lib/solaar/ui/window.py]: battery_level can be NamedInt with value "critical" etc: [/lib/logitech_receiver/hidpp10.py#L102](https://github.com/pwr/Solaar/blob/fa5fba796b87fc1b8f4ce15ac828854ab87a022a/lib/logitech_receiver/hidpp10.py#L102)
Without gettext call here it will not be translated.
Here is translatable strings are defined: [lib/logitech_receiver/i18n.py#L38]( https://github.com/pwr/Solaar/blob/fa5fba796b87fc1b8f4ce15ac828854ab87a022a/lib/logitech_receiver/i18n.py#L38)

**Comment by DJm00n** [comment]
I will prepare new update to this PR.

File: lib/solaar/ui/about.py (hunk 2 above)
Comment (DJm00n) [inline lib/solaar/ui/about.py]: ok

File: lib/solaar/ui/window.py (hunk 9 above)
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: I see, then wouldn't `_(str(battery_level))` be sufficient? (not sure if that gives sufficient context, not very familiar with this)

File: lib/solaar/ui/window.py (hunk 7 above)
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: Since the previous translation was a  set of separate strings, I am leaning towards separate HTML-less strings as well. If you think that mixing markup is fine, then you can do the change anyway I guess

**Comment by DJm00n** [comment]
Rebased and redone PR.

File: lib/solaar/ui/window.py [hunk 10]
```diff
@@ -578,16 +562,15 @@ def _update_receiver_panel(receiver, panel, buttons, full=False):
...
-			panel._count.set_markup(_NANO_RECEIVER_TEXT[0])
-		else:
-			panel._count.set_markup(_NANO_RECEIVER_TEXT[1])
+
+	paired_text = _('No device paired.') if devices_count == 0 else ngettext('%(count)s paired device.', '%(count)s paired devices.', devices_count) % { 'count': devices_count }
+
+	if(receiver.max_devices > 0):
+		paired_text += ngettext('\n\n<small>Up to %(max_count)s device can be paired to this receiver.</small>', '\n\n<small>Up to %(max_count)s devices can be paired to this receiver.</small>', receiver.max_devices) % { 'max_count': receiver.max_devices }
+	elif(devices_count > 0):
+		paired_text += _('\n\n<small>Only one device can be paired to this receiver.</small>')
```
Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: Would it be possible to use `"\n\n<small>%s</small" % _(...)` here to avoid HTML in translations? This would reduce the probability of messing it up in translations.

(here and above in the "Up to ..")

Comment (DJm00n) [inline lib/solaar/ui/window.py]: Gettext manual says that it is ok to include HTML tags in l10n:

> [HTML markup, however, is common enough that it’s probably ok to use in translatable strings. But please bear in mind that the GNU gettext tools don’t verify that the translations are well-formed HTML.](https://www.gnu.org/software/gettext/manual/html_node/Preparing-Strings.html)

Comment (Lekensteyn) [inline lib/solaar/ui/window.py]: This is might be more useful for something like:

    This is <em>something</em> with <strong>inline/strong> formatting.

but there it is merely a container, so I *think* it is better to split it out to reduce errors (avoids malformed HTML). What's your opinion on this?

Comment (DJm00n) [inline lib/solaar/ui/window.py]: Probably you're right here. I updated PR.

**Comment by Lekensteyn** [comment]
LGTM, thanks!
